src/transformation/mesh_intersection/mesh_intersection.rs

```diff
@@ -256,380 +355,358 @@ fn extract_connected_components(
+            return point_count;
+        }
     }
-
-    extract_result(
-        pos12,
-        mesh1,
-        flip1,
-        mesh2,
-        flip2,
-        &spade_handle_to_intersection,
-        &intersection_points,
-        &triangulations1,
-        &triangulations2,
-        new_vertices12,
-        new_indices12,
-    );
 }
 
-fn convert_fid(mesh: &TriMesh, tri: u32, fid: FeatureId) -> FeatureId {
-    match fid {
-        FeatureId::Edge(eid) => {
-            let topology = mesh.topology().unwrap();
-            let half_edge_id = topology.face_half_edges_ids(tri)[eid as usize];
-            let half_edge = &topology.half_edges[half_edge_id as usize];
-            // NOTE: if the twin doesn’t exist, it’s equal to u32::MAX. So the `min` will
-            //       automatically filter it out.
-            FeatureId::Edge(half_edge_id.min(half_edge.twin))
+fn smallest_angle(points: &[Vector3<f64>]) -> f64 {
+    let n = points.len();
+
+    let mut worst_cos = 2.0;
+    for i in 0..points.len() {
+        let d1 = (points[i] - points[(i + 1) % n]).normalize();
+        let d2 = (points[(i + 2) % n] - points[(i + 1) % n]).normalize();
+
+        let cos = d1.dot(&d2);
+
+        if cos < worst_cos {
+            worst_cos = cos.abs();
         }
-        FeatureId::Vertex(vid) => FeatureId::Vertex(mesh.indices()[tri as usize][vid as usize]),
-        FeatureId::Face(_) => FeatureId::Face(tri),
-        FeatureId::Unknown => FeatureId::Unknown,
     }
+
+    worst_cos.acos() * 180. / PI
 }
 
-fn unified_vertex(
-    mesh1: &TriMesh,
-    mesh2: &TriMesh,
-    new_vertices12: &[Point<Real>],
-    pos12: &Isometry<Real>,
-    vid: u32,
-) -> Point<Real> {
-    let base_id2 = mesh1.vertices().len() as u32;
-    let base_id12 = (mesh1.vertices().len() + mesh2.vertices().len()) as u32;
-
-    if vid < base_id2 {
-        mesh1.vertices()[vid as usize]
-    } else if vid < base_id12 {
-        pos12 * mesh2.vertices()[(vid - base_id2) as usize]
-    } else {
-        new_vertices12[(vid - base_id12) as usize]
+fn planar_gram_schmidt(v1: Vector3<f64>, v2: Vector3<f64>) -> (Vector3<f64>, Vector3<f64>) {
+    let u1 = v1;
+    let u2 = v2 - (v2.dot(&u1) / u1.norm_squared()) * u1;
+
+    let e1 = u1.normalize();
+    let e2 = u2.normalize();
+
+    (e1, e2)
+}
+
+fn project_point_to_segment(point: &Vector3<f64>, segment: &[Vector3<f64>; 2]) -> Vector3<f64> {
+    let dir = segment[1] - segment[0];
+    let local = point - segment[0];
+
+    let norm = dir.norm();
+    // restrict the result to the segment portion of the line.
+    let coeff = (dir.dot(&local) / norm).clamp(0., norm);
+
+    segment[0] + coeff * dir.normalize()
+}
+
+/// No matter how smart we are about computing intersections. It is always possible
+/// to create ultra thin triangles when a point lies on an edge of a tirangle. These
+/// are degenerate and need to be terminated with extreme prejudice.
+fn is_triangle_degenerate(
+    triangle: &[Vector3<f64>; 3],
```

Comment:
This should take points since they are the triangle’s vertices.
```suggestion
    triangle: &[Point3<f64>; 3],
```

Reply:
A comment on the three points related to this discussion:

Mathematically, a point is a vector and a vector is a point. Having a distinction between the two increases the total code surface and amount of conversions that maintainers and users have to do. 

For example
```rs
fn smallest_angle(points: &[Point3<f64>]) -> f64 {
    let n = points.len();

    let mut worst_cos = 2.0;
    for i in 0..points.len() {
        let d1 = (points[i].coords - points[(i + 1) % n].coords).normalize();
        let d2 = (points[(i + 2) % n].coords - points[(i + 1) % n].coords).normalize();

        let cos = d1.dot(&d2);

        if cos < worst_cos {
            worst_cos = cos.abs();
        }
    }

    worst_cos.acos() * 180. / PI
}
```

Is adding `.coords` here helping make the code more explicit, or is it adding more effort for the programmer and more complexity for the reader?

Given how common a person needs to compute different kinds of inner products, projections, translations, orthogonalizations... It feels like the `Point` interface and its traits associations produce more difficulties than benefits.

I will change things to comply with the current standard. But I leave these comments to express my humble personal opinion of something that I believe might cause more harm than good.

Reply:
The distinction between points and vectors is mainly to avoid mistakes, like if you wrote `point[i].dot(&point[i + 1])` which would obviously not make sense because that’s taking the dot product of two location in space.

I recommend reading some resources about that distinction. Searching for the distinction between points and vectors yields many results. To site one, from [the university of california](https://ed.iitm.ac.in/~raman/Points-and-Vectors.pdf):
> These objects are often confused when studying the field and frequently students are taught that points and vectors are “essentially” the same. However, they are very different. Namely,
> • A Point has position in space. The only characteristic that distinguishes one point from another is its
position.
> • A Vector has both magnitude and direction, but no fixed position in space.

Regarding your example:


> Is adding .coords here helping make the code more explicit, or is it adding more effort for the programmer and more complexity for the reader?

Adding `.coords` here only illustrates a misunderstanding of the concept of points. Since points are location in spaces, and a vector the displacement between two locations, then the subtraction of two points is a vector. So you can just write `(points[i]- points[(i + 1) % n]).normalize()` because `points[i]- points[(i + 1) % n]` is already a vector in the first place.

Having a distinct type between points and vectors has some additional perks when combined with transformations. For example if `iso` is an isometry, `iso * vector` will only rotate the vector whereas `iso * point` will both rotate and translate it. This prevents a wide range of errors where you would forget to apply the translational part to a vector.

Reply:
The resources that talk about that distinction usually come from the more physics side of mathematics. Many resources will define points and vectors to be the same, for example in wolfram:

> In dimensions greater than or equal to two, points are sometimes considered synonymous with [vectors](https://mathworld.wolfram.com/Vector.html) and so points in [n-dimensional space](https://mathworld.wolfram.com/n-Space.html) are sometimes called [n-vectors](https://mathworld.wolfram.com/n-Vector.html).
https://mathworld.wolfram.com/Point.html

> point[i].dot(&point[i + 1]) which would obviously not make sense

I mean, a contrived counter example is if the points are on a circle and we are looking for the angle between them or if we want to know the winding number of the origin relative to a poly line. 

The only reason why we do things like  `points[i]- points[(i + 1) % n]` is to move objects to the origin to then apply the dot product. So technically speaking `point[i].dot(&point[i + 1])` is the only actually valid operation when taking the dot product, everything else we do is just trying to align our space with the correct origin so that our calculations make sense.

> Since points are location in spaces, and a vector the displacement between two locations,

A vector is not a displacement, a vector is the linear combination of a basis. And since points are represented as coordinates, they are, implicitly, a linear combination of an orthonormal basis.

The idea that vectors are displacements is a specific interpretation in the context of physics, but for example, polynomials and matrices are vector spaces and neither represents a displacement.

This is coming more to a matter of opinion and rather pedantic, as I stated at the beginning. I wish not to distract you from more important subjects. I just wanted to give you my two cents, please disregard this discussion.